torchdp/per_sample_gradient_clip.py

```diff
@@ -171,7 +188,54 @@ def clip(self) -> List[float]:
         if self.stat is not None:
             stats.update(stats.StatType.CLIPPING, "ClippingStats", **self.stat)
             self.stat = {}
-        return threshs
+
+        if virtual:
+            # check if we're adding to an existing accumulator
+            if "clip_threshs" in self.accumulation_state:
+
+                # retain the largest clipping thresholds accross the entire batch
+                # TODO: we could add other strategies for dynamic clipping here,
+                #       as the max threshold for a small mini-batch may be very
+                #       different from the max across the entire effective batch
+                curr_thresh = self.accumulation_state["clip_threshs"]
+                self.accumulation_state["clip_threshs"] = [
+                    max(n1, n2) for (n1, n2) in zip(threshs, curr_thresh)
+                ]
+
+                # keep track of the total batch size
+                self.accumulation_state["batch_size"] += batch_size
+
+            else:
+                # start a new accumulator
+                self.accumulation_state["clip_threshs"] = threshs
+                self.accumulation_state["batch_size"] = batch_size
+
+        return threshs, batch_size
+
+    def finalize_batch(self) -> (List[float], int):
+        """
+        Averages the clipped gradients aggregated over multiple mini-batches and
+        stores them in the .grad field
+        """
+
+        batch_size = self.accumulation_state["batch_size"]
+        # now that we know the full batch size, we can average the gradients
+        for _, p in self.named_params:
+            acc_grad = p.summed_grad
+            p.grad = acc_grad / batch_size
```

Comment:
same here,  from the looks of it we can just discard p.summed-grad here

Reply:
This happens in the call to erase_virtual_batch 2 lines below.
I extracted this into its own methos so that it can also be called in .zero_grad(), to address the same concern as above, namely that we could finish a training epoch with an incomplete effective batch.